File: google-app-scripts/sheethub/test/marshal/ruby.test.ts
import {dump, load, MAJOR_VERSION, MINOR_VERSION, OFFSET, TYPE_FIXNUM, TYPE_STRING} from "../../src/marshal/ruby";
import {describe, expect, test} from "vitest";

describe("dump", () => {
    const majorVersionCode = String.fromCharCode(MAJOR_VERSION);
    const minorVersionCode = String.fromCharCode(MINOR_VERSION);

    describe("fixnum", () => {
        const typeFixnumCode = String.fromCharCode(TYPE_FIXNUM);

        test("zero", () => {
            const n = 0;

            expect(dump(n)).toStrictEqual(`${majorVersionCode}${minorVersionCode}${typeFixnumCode}${String.fromCharCode(0)}`);
        });

        test("overflow", () => {
            const n = 2 ** 31;

            expect(() => dump(n)).toThrow("long too big to dump");
        });

        test("short", () => {
            const n = 1;

            expect(dump(n)).toStrictEqual(`${majorVersionCode}${minorVersionCode}${typeFixnumCode}${String.fromCharCode(n + OFFSET)}`);
        });

        test("long", () => {
            const n = 2 ** 31 - 1;

            const long = new Uint8Array(4);
            long[0] = n & 0xff;
            long[1] = (n >> 8) & 0xff;
            long[2] = (n >> 16) & 0xff;
            long[3] = (n >> 24) & 0xff;
            const index = long.length - Array.from(long).reverse().findIndex((value) => value !== 0);

            expect(dump(n)).toStrictEqual(`${majorVersionCode}${minorVersionCode}${typeFixnumCode}${String.fromCharCode(index)}${Array.from(long.subarray(0, index)).map((value) => String.fromCharCode(value)).join("")}`);
        });
    });

    describe("string", () => {
        const typeStringCode = String.fromCharCode(TYPE_STRING);

        test("short", () => {
            const s = "test";

            const length = String.fromCharCode(s.length + 5);

            expect(dump(s)).toStrictEqual(`${majorVersionCode}${minorVersionCode}${typeStringCode}${length}${s}`);
        });

        test("long", () => {
            const s = "test".repeat(100);

            const long = new Uint8Array(4);
            long[0] = s.length & 0xff;
            long[1] = (s.length >> 8) & 0xff;
            long[2] = (s.length >> 16) & 0xff;
            long[3] = (s.length >> 24) & 0xff;
            const index = long.length - Array.from(long).reverse().findIndex((value) => value !== 0);
            const length = Array.from(long.subarray(0, index)).map((value) => String.fromCharCode(value)).join("");

            expect(dump(s)).toStrictEqual(`${majorVersionCode}${minorVersionCode}${typeStringCode}${String.fromCharCode(index)}${length}${s}`);
        });

        test("multibyte", () => {
            const s = "テスト";

            const length = String.fromCharCode(utf8ByteLength("テ".codePointAt(0)!) + utf8ByteLength("ス".codePointAt(0)!) + utf8ByteLength("ト".codePointAt(0)!) + 5);

            expect(dump(s)).toStrictEqual(`${majorVersionCode}${minorVersionCode}${typeStringCode}${length}${s}`);
        });
    });
});

describe("load", () => {
    const majorVersionCode = String.fromCharCode(MAJOR_VERSION);
    const minorVersionCode = String.fromCharCode(MINOR_VERSION);

    describe("fixnum", () => {
        const typeFixnumCode = String.fromCharCode(TYPE_FIXNUM);

        test("zero", () => {
            const n = 0;

            expect(load(`${majorVersionCode}${minorVersionCode}${typeFixnumCode}${String.fromCharCode(0)}`)).toStrictEqual(n);
        });

        test("short", () => {
            const n = 1;

            expect(load(`${majorVersionCode}${minorVersionCode}${typeFixnumCode}${String.fromCharCode(n + OFFSET)}`)).toStrictEqual(n);
        });

        test("long", () => {
            const n = 2 ** 31 - 1;

            const long = new Uint8Array(4);
            long[0] = n & 0xff;
            long[1] = (n >> 8) & 0xff;
            long[2] = (n >> 16) & 0xff;
            long[3] = (n >> 24) & 0xff;
            const index = long.length - Array.from(long).reverse().findIndex((value) => value !== 0);

            expect(load(`${majorVersionCode}${minorVersionCode}${typeFixnumCode}${String.fromCharCode(index)}${Array.from(long.subarray(0, index)).map((value) => String.fromCharCode(value)).join("")}`)).toStrictEqual(n);
        });
    });

    describe("string", () => {
        const typeStringCode = String.fromCharCode(TYPE_STRING);

        test("short", () => {
            const s = "test";

            const length = s.length + 5;

            expect(load(`${majorVersionCode}${minorVersionCode}${typeStringCode}${String.fromCharCode(length)}${s}`)).toStrictEqual(s);
        });

        test("long", () => {
            const s = "test".repeat(100);

            const long = new Uint8Array(4);
            long[0] = s.length & 0xff;
            long[1] = (s.length >> 8) & 0xff;
            long[2] = (s.length >> 16) & 0xff;
            long[3] = (s.length >> 24) & 0xff;
            const index = long.length - Array.from(long).reverse().findIndex((value) => value !== 0);
            const length = Array.from(long.subarray(0, index)).map((value) => String.fromCharCode(value)).join("");

            expect(load(`${majorVersionCode}${minorVersionCode}${typeStringCode}${String.fromCharCode(index)}${length}${s}`)).toStrictEqual(s);
        });

        test("multibyte", () => {
            const s = "テスト";

            const length = utf8ByteLength("テ".codePointAt(0)!) + utf8ByteLength("ス".codePointAt(0)!) + utf8ByteLength("ト".codePointAt(0)!) + 5;

            expect(load(`${majorVersionCode}${minorVersionCode}${typeStringCode}${String.fromCharCode(length)}${s}`)).toStrictEqual(s);
        });
    });
});

const SurrogateMin = 0xd800;
const SurrogateMax = 0xdfff;
const Rune1Max = (1 << 7) - 1;
const Rune2Max = (1 << 11) - 1;
const Rune3Max = (1 << 16) - 1;
const MaxRune = 0x10ffff;

const utf8ByteLength = (rune: number): number => {
    if (rune < 0) {
        return -1;
    } else if (rune <= Rune1Max) {
        return 1;
    } else if (rune <= Rune2Max) {
        return 2;
    } else if (SurrogateMin <= rune && rune <= SurrogateMax) {
        return -1;
    } else if (rune <= Rune3Max) {
        return 3;
    } else if (rune <= MaxRune) {
        return 4;
    } else {
        return -1;
    }
}

File: armyknife/internal/marshal/ruby/ruby_test.go
package ruby_test

import (
	"armyknife/internal/marshal/ruby"
	"fmt"
	"math"
	"strings"
	"testing"
)

func TestDump(t *testing.T) {
	majorVersionCode := string(rune(ruby.MajorVersion))
	minorVersionCode := string(rune(ruby.MinorVersion))

	t.Run("fixnum", func(t *testing.T) {
		typeFixnumCode := string(ruby.TypeFixnum)

		t.Run("zero", func(t *testing.T) {
			n := 0

			expected := fmt.Sprintf("%s%s%s%s", majorVersionCode, minorVersionCode, typeFixnumCode, string(rune(0)))
			result, err := ruby.Dump(n)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != expected {
				t.Errorf("Expected %q, got %q", expected, result)
			}
		})

		t.Run("overflow", func(t *testing.T) {
			n := int(math.Pow(2, 31))

			if _, err := ruby.Dump(n); err == nil {
				t.Error("Expected error, got nil")
			}
		})

		t.Run("short", func(t *testing.T) {
			n := 1

			expected := fmt.Sprintf("%s%s%s%s", majorVersionCode, minorVersionCode, typeFixnumCode, string(rune(n+ruby.Offset)))
			result, err := ruby.Dump(n)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != expected {
				t.Errorf("Expected %q, got %q", expected, result)
			}
		})

		t.Run("long", func(t *testing.T) {
			n := math.MaxInt32

			long := make([]byte, 4)
			long[0] = byte(n & 0xff)
			long[1] = byte((n >> 8) & 0xff)
			long[2] = byte((n >> 16) & 0xff)
			long[3] = byte((n >> 24) & 0xff)

			index := len(long)
			for i := len(long) - 1; i >= 0; i-- {
				if long[i] != 0 {
					index = i + 1
					break
				}
			}

			var s string
			for i := 0; i < index; i++ {
				s += string(rune(long[i]))
			}

			expected := fmt.Sprintf("%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeFixnumCode,
				string(rune(index)),
				s,
			)
			result, err := ruby.Dump(n)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != expected {
				t.Errorf("Expected %q, got %q", expected, result)
			}
		})
	})

	t.Run("string", func(t *testing.T) {
		typeStringCode := string(rune(ruby.TypeString))

		t.Run("short", func(t *testing.T) {
			s := "test"

			length := string(rune(len(s) + 5))

			expected := fmt.Sprintf("%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeStringCode,
				length,
				s,
			)
			result, err := ruby.Dump(s)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != expected {
				t.Errorf("Expected %q, got %q", expected, result)
			}
		})

		t.Run("long", func(t *testing.T) {
			s := strings.Repeat("test", 100)

			long := make([]byte, 4)
			long[0] = byte(len(s) & 0xff)
			long[1] = byte((len(s) >> 8) & 0xff)
			long[2] = byte((len(s) >> 16) & 0xff)
			long[3] = byte((len(s) >> 24) & 0xff)

			index := len(long)
			for i := len(long) - 1; i >= 0; i-- {
				if long[i] != 0 {
					index = i + 1
					break
				}
			}

			var length string
			for i := 0; i < index; i++ {
				length += string(rune(long[i]))
			}

			expected := fmt.Sprintf("%s%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeStringCode,
				string(rune(index)),
				length,
				s,
			)
			result, err := ruby.Dump(s)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != expected {
				t.Errorf("Expected %q, got %q", expected, result)
			}
		})

		t.Run("multibyte", func(t *testing.T) {
			s := "テスト"

			length := string(rune(len(s) + 5))

			expected := fmt.Sprintf("%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeStringCode,
				length,
				s,
			)
			result, err := ruby.Dump(s)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != expected {
				t.Errorf("Expected %q, got %q", expected, result)
			}
		})
	})
}

func TestLoad(t *testing.T) {
	majorVersionCode := string(rune(ruby.MajorVersion))
	minorVersionCode := string(rune(ruby.MinorVersion))

	t.Run("fixnum", func(t *testing.T) {
		typeFixnumCode := string(rune(ruby.TypeFixnum))

		t.Run("zero", func(t *testing.T) {
			n := 0

			result, err := ruby.Load(fmt.Sprintf("%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeFixnumCode,
				string(rune(0)),
			))
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != n {
				t.Errorf("Expected 0, got %v", result)
			}
		})

		t.Run("short", func(t *testing.T) {
			n := 1

			result, err := ruby.Load(fmt.Sprintf("%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeFixnumCode,
				string(rune(n+ruby.Offset)),
			))
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != n {
				t.Errorf("Expected %d, got %v", n, result)
			}
		})

		t.Run("long", func(t *testing.T) {
			n := math.MaxInt32

			long := make([]byte, 4)
			long[0] = byte(n & 0xff)
			long[1] = byte((n >> 8) & 0xff)
			long[2] = byte((n >> 16) & 0xff)
			long[3] = byte((n >> 24) & 0xff)

			index := len(long)
			for i := len(long) - 1; i >= 0; i-- {
				if long[i] != 0 {
					index = i + 1
					break
				}
			}

			var length string
			for i := 0; i < index; i++ {
				length += string(rune(long[i]))
			}

			result, err := ruby.Load(fmt.Sprintf("%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeFixnumCode,
				string(rune(index)),
				length,
			))
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != n {
				t.Errorf("Expected %d, got %v", n, result)
			}
		})
	})

	t.Run("string", func(t *testing.T) {
		typeStringCode := string(rune(ruby.TypeString))

		t.Run("short", func(t *testing.T) {
			s := "test"

			result, err := ruby.Load(fmt.Sprintf("%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeStringCode,
				string(rune(len(s)+5)),
				s,
			))
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != s {
				t.Errorf("Expected %q, got %q", s, result)
			}
		})

		t.Run("long", func(t *testing.T) {
			s := strings.Repeat("test", 100)

			long := make([]byte, 4)
			long[0] = byte(len(s) & 0xff)
			long[1] = byte((len(s) >> 8) & 0xff)
			long[2] = byte((len(s) >> 16) & 0xff)
			long[3] = byte((len(s) >> 24) & 0xff)

			index := len(long)
			for i := len(long) - 1; i >= 0; i-- {
				if long[i] != 0 {
					index = i + 1
					break
				}
			}

			var length string
			for i := 0; i < index; i++ {
				length += string(rune(long[i]))
			}

			result, err := ruby.Load(fmt.Sprintf("%s%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeStringCode,
				string(rune(index)),
				length,
				s,
			))
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != s {
				t.Errorf("Expected %q, got %q", s, result)
			}
		})

		t.Run("multibyte", func(t *testing.T) {
			s := "テスト"

			result, err := ruby.Load(fmt.Sprintf("%s%s%s%s%s",
				majorVersionCode,
				minorVersionCode,
				typeStringCode,
				string(rune(len(s)+5)),
				s,
			))
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result != s {
				t.Errorf("Expected %q, got %q", s, result)
			}
		})
	})
}
